contracts/examples/esdt-transfer-with-fee/src/esdt_transfer_with_fee.rs

```diff
@@ -0,0 +1,133 @@
+#![no_std]
+
+mod fee;
+use fee::*;
+
+elrond_wasm::imports!();
+#[elrond_wasm::contract]
+pub trait EsdtTransferWithFee {
+    #[init]
+    fn init(&self) {}
+
+    #[only_owner]
+    #[endpoint(setExactValueFee)]
+    fn set_exact_value_fee(&self, fee: EsdtTokenPayment<Self::Api>, token: TokenIdentifier) {
+        self.token_fee(&token).set(Fee::ExactValue(fee));
+    }
+
+    #[only_owner]
+    #[endpoint(setPercentageFee)]
+    fn set_percentage_fee(&self, fee: u32, token: TokenIdentifier) {
+        self.token_fee(&token).set(Fee::Percentage(fee));
+    }
+
+    #[only_owner]
+    #[endpoint(claimFees)]
+    fn claim_fees(&self) {
+        let paid_fees = self.paid_fees();
+        require!(!paid_fees.is_empty(), "There is nothing to claim");
+        let mut fees = ManagedVec::new();
+        for ((token, nonce), amount) in self.paid_fees().iter() {
+            fees.push(EsdtTokenPayment::new(token, nonce, amount))
+        }
+        self.paid_fees().clear();
+
+        let caller = self.blockchain().get_caller();
+        self.send().direct_multi(&caller, &fees, &[]);
+    }
+
+    #[payable("*")]
+    #[endpoint]
+    fn transfer(&self, address: ManagedAddress) {
+        let payments = self.call_value().all_esdt_transfers();
+        let mut new_payments = ManagedVec::new();
+
+        let mut payments_iter = payments.iter();
+        while let Some(payment) = payments_iter.next() {
+            let fee_type = self.get_fee_type(&payment);
+            match &fee_type {
+                Fee::ExactValue(fee) => {
+                    let next_payment = payments_iter
+                        .next()
+                        .unwrap_or_else(|| sc_panic!("Fee payment missing"));
+                    require!(
+                        next_payment.token_identifier == fee.token_identifier
+                            && next_payment.token_nonce == fee.token_nonce,
+                        "Fee payment missing"
+                    );
+                    require!(
+                        next_payment.amount == fee.amount,
+                        "Insufficient payments for covering fees"
+                    );
+
+                    self.get_payment_after_fees(fee_type, &next_payment);
+                    new_payments.push(payment);
+                },
+                Fee::Percentage(_) => {
+                    self.get_payment_after_fees(fee_type, &payment);
+                    new_payments.push(payment);
```

Comment:
Hm, it seems like you don't cut anything from the original payment here.

Reply:
So you should probably either use the payment returned from `get_payment_after_fees`, or give a mutable reference to the function so it modifies that directly.